### PR TITLE
Update time over SIM

### PR DESCRIPTION
**Pull request by akoen**

### CODE REVIEW

File: avionics/envs/x86/lib/HAL/time.cpp [hunk 1]
```diff
@@ -35,11 +30,15 @@ void initialSystem() {
         std::chrono::time_point_cast<ms>(std::chrono::steady_clock::now());
 }
 
+/**
+ * @brief Retuns the current time in milliseconds relative to startup_t_point.
+ */
 t_point now_ms() {
```
Comment (akoen) [inline avionics/envs/x86/lib/HAL/time.cpp:38]: @jesseli2002 I'd appreciate it if you could give this function a good look over: it seems to work well during bench testing but still confuses me somewhat.

File: avionics/envs/x86/include/stdio_controller.hpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <atomic>
+#include <cassert>
```
Comment (SilentSys) [inline avionics/envs/x86/include/stdio_controller.hpp]:
```suggestion
```
Doesn't look like this is used anywhere

File: avionics/envs/x86/include/stdio_controller.hpp [hunk 3]
```diff
@@ -12,10 +13,13 @@
 #include <unordered_map> // for hash map
 #include <vector>
 
+#include "HAL/time.h"
```
Comment (SilentSys) [inline avionics/envs/x86/include/stdio_controller.hpp]: Is this used in here?

File: avionics/envs/x86/lib/HAL/time.cpp [hunk 4]
```diff
@@ -1,31 +1,26 @@
...
-}
-
 /**
  * @brief Sleep for some number of milliseconds.
  * @param t Duration to sleep for, in milliseconds.
  */
-void sleep_ms(uint32_t t) { sleep(ms(t)); }
+void sleep_ms(uint32_t t) {
+    if (t >= (std::numeric_limits<uint32_t>::max() / 1000))
+        throw "The given sleep length will cause overflow.";
```
Comment (SilentSys) [inline avionics/envs/x86/lib/HAL/time.cpp]: 👍👍👍

**Review by SilentSys — CHANGES_REQUESTED** [review]
LGTM!

You'll need to update `required_ubcgs.txt` in the root directory to contain the commit hash from https://github.com/UBC-Rocket/UBCRocketGroundStation/pull/46 for the CI tests to pass.

**Comment by akoen** [comment]
> You'll need to update `required_ubcgs.txt` in the root directory to contain the commit hash from [UBC-Rocket/UBCRocketGroundStation#46](https://github.com/UBC-Rocket/UBCRocketGroundStation/pull/46) for the CI tests to pass.

Apologies I'm a little unclear on this part. Do I replace the existing hash and with which commit hash do I replace it with?

**Comment by SilentSys** [comment]
> > You'll need to update `required_ubcgs.txt` in the root directory to contain the commit hash from [UBC-Rocket/UBCRocketGroundStation#46](https://github.com/UBC-Rocket/UBCRocketGroundStation/pull/46) for the CI tests to pass.
> 
> Apologies I'm a little unclear on this part. Do I replace the existing hash and with which commit hash do I replace it with?

Sorry, its not yet documented anywhere. Yeah you replace the one that's currently in there with the hash of whichever commit you want to test against, probably your latest commit in that PR:
```
commit 3b9c808301786cdb73b08b7771ba60bbdc4b9e7e (HEAD -> gs-clock, origin/gs-clock)
Author: akoen <arw.koen@gmail.com>
Date:   Thu Nov 26 09:40:37 2020 -0800

    Review changes
```

**Comment by SilentSys** [comment]
[Method for synchronizing time over SIM](https://trello.com/c/WnWE6PZb/245-method-for-synchronizing-time-over-sim)

**Comment by akoen** [comment]
Alright integrations tests should pass now.

**Comment by SilentSys** [comment]
Had to fix a merge conflict for the CI to trigger. Yup passes now! And damn, the CI job completed twice as fast since we don't need to sleep actual time 😋. Love it.

Had to restart the CI a few times for it to pass. Had the same issue on one of my PR's. Pretty sure its Travis CI is just having some technical issues... Hopefully its not an instability on our side...

**Review by SilentSys — APPROVED** [review]

File: avionics/envs/x86/lib/HAL/time.cpp (hunk 4 above)
Comment (jesseli2002) [inline avionics/envs/x86/lib/HAL/time.cpp]: Can't say I know too much about C++ exceptions, but my understanding is that you should throw things related to std::exception, and I'm not sure how string literals will behave with stack unwinding and potential undefined behaviour - see https://stackoverflow.com/questions/27179011/c-exception-throw-a-string and https://stackoverflow.com/questions/27179011/c-exception-throw-a-string

File: avionics/envs/x86/lib/HAL/time.cpp (hunk 1 above)
Comment (jesseli2002) [reply avionics/envs/x86/lib/HAL/time.cpp:38]: Seems fine.

**Review by jesseli2002 — COMMENTED** [review]
Looks mostly fine. It might be worth to add a comment in avionics.cpp that the sleeping call is necessary to actually move time forward.

File: avionics/envs/x86/lib/HAL/time.cpp (hunk 1 above)
Comment (jesseli2002) [reply avionics/envs/x86/lib/HAL/time.cpp:38]: Probably best to put in a comment noting that passing 0 ends up having no effect on the resulting time, though - it's too clever to figure out if you're tired. Maybe slip in a note in the confluence spec as well.

**Review by jesseli2002 — APPROVED** [review]

**Pull request merged**